aten/src/ATen/native/nested/cuda/NestedTensorTransformerUtils.cpp

```diff
@@ -449,6 +472,100 @@ sdpa_nested_preprocessing(
+    const at::Tensor& key,
+    const at::Tensor& value,
+    const at::Tensor& out,
+    const Tensor& cumulative_sequence_length_q,
+    const Tensor& cumulative_sequence_length_kv,
+    const int64_t max_seqlen_batch_q,
+    const int64_t max_seqlen_batch_kv) {
+  const int64_t q_batch_size = query.size(0);
+  const int64_t k_batch_size = key.size(0);
+
+  const int64_t v_batch_size = value.size(0);
+
+  const int64_t q_num_heads = query.size(1);
+  const int64_t k_num_heads = key.size(1);
+  const int64_t v_num_heads = value.size(1);
+
+  if (!(q_batch_size == k_batch_size && q_batch_size == v_batch_size) ||
+      !(q_num_heads == k_num_heads && k_num_heads == v_num_heads)) {
+        TORCH_CHECK(false, "Broadcasted NestedTensor inputs is currently not supported for backwards.");
+  }
+
+  const int64_t num_heads = query.size(1);
+  const int64_t head_dim_qk = query.size(3);
+  const int64_t head_dim_v = value.size(3);
+
+  Tensor q_t = query.transpose(1, 2);
+  Tensor k_t = key.transpose(1, 2);
+  Tensor v_t = value.transpose(1, 2);
+  Tensor grad_out_t = grad_out_.transpose(1, 2);
+  Tensor out_t = out.transpose(1, 2);
+
+  const int64_t Nnz_q = get_nnz(q_t);
+  const int64_t Nnz_kv = get_nnz(k_t);
+
+  Tensor query_buffer_reshaped;
+  Tensor key_buffer_reshaped;
+  Tensor value_buffer_reshaped;
+  Tensor grad_out_buffer_reshaped;
+  Tensor output_buffer_reshaped;
+
+  const auto* query_impl = get_nested_tensor_impl(q_t);
+  const auto* key_impl = get_nested_tensor_impl(k_t);
+  const auto* value_impl = get_nested_tensor_impl(v_t);
+  const auto* grad_out_impl = get_nested_tensor_impl(grad_out_t);
+  const auto* out_impl = get_nested_tensor_impl(out_t);
+
+  // If the physical layout of the NestedTensor's storage
+  // is not: batch, {seq_len}, num_heads, head_dim then we need
+  // to call contiguous
+  if (!q_t.is_contiguous() && !is_safe_to_get_storage_as_tensor(query_impl)) {
+    q_t = q_t.contiguous();
+  }
+  if (!k_t.is_contiguous() && !is_safe_to_get_storage_as_tensor(key_impl)) {
+    k_t = k_t.contiguous();
+  }
+  if (!v_t.is_contiguous() && !is_safe_to_get_storage_as_tensor(value_impl)) {
+    v_t = v_t.contiguous();
+  }
+  if (!grad_out_t.is_contiguous() && !is_safe_to_get_storage_as_tensor(grad_out_impl)) {
+    grad_out_t = grad_out_t.contiguous();
+  }
+  if (!out_t.is_contiguous() && !is_safe_to_get_storage_as_tensor(out_impl)) {
+    out_t = out_t.contiguous();
+  }
+
+  query_buffer_reshaped = view_as_dense(q_t, Nnz_q, num_heads, head_dim_qk);
+  key_buffer_reshaped = view_as_dense(k_t, Nnz_kv, num_heads, head_dim_qk);
+  value_buffer_reshaped = view_as_dense(v_t, Nnz_kv, num_heads, head_dim_v);
+
+ grad_out_buffer_reshaped =
+      view_as_dense(grad_out_t, Nnz_q, num_heads, head_dim_v);
+  output_buffer_reshaped = view_as_dense(out_t, Nnz_q, num_heads, head_dim_v);
+
+  auto output_shape = get_nested_sizes(q_t).clone();
+  if (head_dim_v != head_dim_qk) {
+    output_shape.select(1, -1).fill_(head_dim_v);
+  }
```

Comment:
The last dim for k and v need not be the same for SDPA, and efficient attention supports this. Since we are only adding train support for flash right now it is not used but if we applied the same for mem eff it would be useful